c++/src/kj/compat/http.c++

```diff
@@ -4093,6 +4093,103 @@ private:
+  // is permanently errored and existing pending calls (reads and writes) are canceled.
+public:
+  AsyncIoStreamWithDeferredInput(
+      kj::Own<kj::AsyncIoStream> inner,
+      kj::Promise<HttpInputStreamImpl::ReleasedBuffer> guard)
+      : inner(kj::mv(inner)),
+        guard(handleGuard(kj::mv(guard))) {}
+
+  // AsyncInputStream
+  Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return guard.addBranch().then([this, buffer, minBytes, maxBytes] {
+      return inner->tryRead(buffer, minBytes, maxBytes);
+    });
+  }
+
+  Maybe<uint64_t> tryGetLength() override {
+    return nullptr;
+  }
+
+  kj::Promise<uint64_t> pumpTo(AsyncOutputStream& output, uint64_t amount) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return guard.addBranch().then([this, &output, amount] {
+      return inner->pumpTo(output, amount);
+    });
+  }
+
+  // AsyncOutputStream
+
+  void shutdownWrite() override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      throw *exception;
+    }
+    inner->shutdownWrite();
+  }
+
+  kj::Maybe<kj::Promise<uint64_t>> tryPumpFrom(AsyncInputStream& input, uint64_t amount) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::Maybe<kj::Promise<uint64_t>>(kj::cp(*exception));
+    }
+    return canceler.wrap(input.pumpTo(*inner, amount));
+  }
+
+  Promise<void> write(const void* buffer, size_t size) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return canceler.wrap(inner->write(buffer, size));
+  }
+
+  Promise<void> write(ArrayPtr<const ArrayPtr<const byte>> pieces) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return canceler.wrap(inner->write(pieces));
+  }
+
+  Promise<void> whenWriteDisconnected() override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return canceler.wrap(inner->whenWriteDisconnected());
+  }
+
+private:
+  kj::Own<kj::AsyncIoStream> inner;
+  kj::ForkedPromise<void> guard;
+  kj::Maybe<kj::Exception> permanentException;
+  kj::Canceler canceler;
+
+  kj::ForkedPromise<void> handleGuard(kj::Promise<HttpInputStreamImpl::ReleasedBuffer> guard) {
+    return guard.then([this](HttpInputStreamImpl::ReleasedBuffer buffer) {
+      inner = heap<AsyncIoStreamWithInitialBuffer>(
+          kj::mv(inner),
+          kj::mv(buffer.buffer), buffer.leftover);
+    }, [this](kj::Exception&& exception) {
+      // Set the permanent exception to catch new calls...
+      permanentException = kj::cp(exception);
+      // Trigger the canceler to interupt existing calls...
+      canceler.cancel(kj::cp(exception));
+      // Return the exception to error out the write guard...
+      return kj::mv(exception);
```

Comment:
Since the success continuation returns void, the return value of the error continuation will be ignored, i.e. this exception will get swallowed. This is interesting, I don't think I've seen this before but it seems like a pretty big gotcha.

You need to do: `kj::throwRecoverableException(kj::mv(exception))`